include/rmm/mr/device/callback_memory_resource.hpp

```diff
@@ -0,0 +1,121 @@
+ *   `device_memory_resource::deallocate`.
+ *
+ * * Additional context to the callback may be passed via `arg` that is provided at
+ *   construction of the `callback_memory_resource`.
+ */
+using deallocate_callback_t = std::function<void(void*, std::size_t, cuda_stream_view, void*)>;
+
+/**
+ * @brief A device memory resource that uses the provided callbacks for memory allocation
+ * and deallocation.
+ */
+class callback_memory_resource final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new callback memory resource.
+   *
+   * Constructs a callback memory resource that uses the user-provided callbacks
+   * `allocate_callback` for allocation and `deallocate_callback` for deallocation.
+   *
+   * @param allocate_callback The callback function used for allocation
+   * @param deallocate_callback The callback function used for deallocation
+   * @param allocate_callback_arg Additional context passed to `allocate_callback`.
+   * It is the caller's responsibility to maintain the lifetime of the pointed-to data
+   * for the duration of the lifetime of the `callback_memory_resource`.
+   * @param deallocate_callback_arg Additional context passed to `deallocate_callback`.
+   * It is the caller's responsibility to maintain the lifetime of the pointed-to data
+   * for the duration of the lifetime of the `callback_memory_resource`.
+   */
+  callback_memory_resource(allocate_callback_t allocate_callback,
+                           deallocate_callback_t deallocate_callback,
+                           void* allocate_callback_arg,
+                           void* deallocate_callback_arg) noexcept
+    : allocate_callback_(allocate_callback),
+      deallocate_callback_(deallocate_callback),
+      allocate_callback_arg_(allocate_callback_arg),
+      deallocate_callback_arg_(deallocate_callback_arg)
+  {
+  }
+
+  callback_memory_resource()                                = delete;
+  ~callback_memory_resource() override                      = default;
+  callback_memory_resource(callback_memory_resource const&) = delete;
+  callback_memory_resource& operator=(callback_memory_resource const&) = delete;
+  callback_memory_resource(callback_memory_resource&&) noexcept        = default;
+  callback_memory_resource& operator=(callback_memory_resource&&) noexcept = default;
+
+ private:
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    return allocate_callback_(bytes, stream, allocate_callback_arg_);
+  }
+
+  void do_deallocate(void* ptr, std::size_t bytes, cuda_stream_view stream) override
+  {
+    deallocate_callback_(ptr, bytes, stream, deallocate_callback_arg_);
+  }
+
+  [[nodiscard]] std::pair<std::size_t, std::size_t> do_get_mem_info(cuda_stream_view) const override
+  {
+    throw std::runtime_error("cannot get free / total memory");
+  }
+
+  [[nodiscard]] virtual bool supports_streams() const noexcept { return false; }
```

Comment:
This doesn't support streams? Is it because you can't expect the callback to properly handle them?

Reply:
I don't think there is any right answer here; the callback may handle streams or it may not. @jrhemstad thoughts?

Reply:
Technically the answer should be provided by another callback function.

In truth, the `supports_streams` function is silly and I should have never added it. It will be removed in the new interface. You can just make this always return true and it doesn't really matter.